### PR TITLE
Pages: integrate with invenio-administration [REQUIRES inveniosoftware/invenio-administration#160]

### PR DESCRIPTION
**Pull request by Pineirin**
closes: #66 
requires: https://github.com/inveniosoftware/invenio-administration/pull/160

Screenshots respectively in Desktop, Laptop and Tablet resolutions.

Search view:
Screenshot from 2023-02-28 10-55-01
Screenshot from 2023-02-28 10-55-11
Screenshot from 2023-02-28 10-55-21

Details view:
Screenshot from 2023-02-28 11-14-19
Screenshot from 2023-02-28 11-14-27
Screenshot from 2023-02-28 11-14-36

Edit view:
Screenshot from 2023-02-28 10-56-44
Screenshot from 2023-02-28 10-56-51
Screenshot from 2023-02-28 10-56-36

### CODE REVIEW

File: invenio_pages/templates/invenio_pages/administration/page_details.html [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{#
+  Copyright (C) 2022 CERN.
+
+  Invenio Pages is free software; you can redistribute it and/or modify it
+  under the terms of the MIT License; see LICENSE file for more details.
+#}
+
+{% extends "invenio_administration/details.html" %}
```
Comment (anikachurilova) [inline invenio_pages/templates/invenio_pages/administration/page_details.html:8]: It's not necessary to override the `invenio-administration` details page if we don't need to change/add anything. It will automatically use the `invenio_administration/details.html` file if we don't specify any `template` in config file. Same for the search page.

Comment (Pineirin) [reply invenio_pages/templates/invenio_pages/administration/page_details.html:8]: In my case I'm needing to override the invenio-administration details and search views, as I need to make adjustments to the jsx for it to look nice.

File: invenio_pages/records/errors.py [hunk 2]
```diff
@@ -12,23 +12,15 @@ class PageNotFoundError(Exception):
...
     def __init__(self, identifier):
-        """Constructor."""
-        self.identifier = identifier
-
-    @property
-    def description(self):
-        """Exception's description."""
-        return f"The page identified by {self.identifier} cannot be found."
+        """Initialise error."""
+        super().__init__(f"The page identified by {identifier} cannot be found.")
```
Comment (ppanero) [inline invenio_pages/records/errors.py:16]: question: what is the reason behind this change? I'm not saying is not correct, however it does change the exception's API. Before the message was stored in the `description` property, while now is in the exeption's `args`.

File: invenio_pages/services/results.py [hunk 3]
```diff
@@ -75,6 +70,23 @@ def _results(self):
             else self._page_results
         )
 
+    def to_dict(self):
```
Comment (ppanero) [inline invenio_pages/services/results.py:73]: moderate: why can't we use the parent class function? If I understand correctly the only difference is that we force disallow aggregations and fields' expansion, what was the reason behind this decision? Otherwise, I think we could live with the parent one.

File: invenio_pages/templates/invenio_pages/administration/page-search.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{#
+  Copyright (C) 2022 CERN.
```
Comment (ppanero) [inline invenio_pages/templates/invenio_pages/administration/page-search.html:2]: Remember to update the high end of the ranges in the files you modify.
```suggestion
  Copyright (C) 2023 CERN.
```

File: invenio_pages/views.py [hunk 5]
```diff
@@ -61,11 +61,11 @@ def render_page(path):
     :returns: The rendered template.
     """
     try:
-        page = current_pages_service.read_by_url(g.identity, request.path).to_obj
+        page = current_pages_service.read_by_url(g.identity, request.path).to_dict()
```
Comment (ppanero) [inline invenio_pages/views.py:64]: minor: why would obj not work? maybe we have a serialization problem that needs fixing instead of changing this?

**Review by ppanero — COMMENTED** [review]
I mainly have questions not comments, so bear with me 😅 I reviewed mostly the code of it, I would like someone more proficient in how invenio-administration works to review

File: invenio_pages/views.py (hunk 5 above)
Comment (Pineirin) [reply invenio_pages/views.py:64]: to_obj was a hackish property I added to return the record.

File: invenio_pages/services/results.py (hunk 3 above)
Comment (Pineirin) [reply invenio_pages/services/results.py:73]: This was a weird issue found by @anikachurilova by which the sortBy params of the parent to_dict() were breaking the search.

File: invenio_pages/records/errors.py (hunk 2 above)
Comment (Pineirin) [reply invenio_pages/records/errors.py:16]: I got asked to change it in a previous comment. I used as example the errors.py of invenio_administration.

File: invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultItem.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      displayEdit,
+      displayDelete,
+      actions,
+      idKeyPath,
+      resourceSchema,
+      listUIEndpoint,
+    } = this.props;
+    const resourceHasActions =
+      displayEdit || displayDelete || !isEmpty(actions);
+    result.content = _truncate(result.content, { length: 250 });
```
Comment (alejandromumo) [inline invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultItem.js:40]: question: the input prop is being mutated by the component. So for a result item with length > 250, does this code truncate the content and never allows the user to read its full value?

File: invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultItem.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            >
+              {index === 0 && (
+                <a
+                  href={AdminUIRoutes.detailsView(
+                    listUIEndpoint,
+                    result,
+                    idKeyPath
+                  )}
+                >
+                  <Formatter
```
Comment (alejandromumo) [inline invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultItem.js:59]: Is this formatter needed here?

I see there are three [formatters](https://github.dev/inveniosoftware/invenio-administration/blob/e3a8cbf68260a9f0ef68776375a119c197a34a92/invenio_administration/assets/semantic-ui/js/invenio_administration/src/components/Formatter.js#L14) (`date`, `datetime` and `bool`) but `pages` uses the `url` as the first column.

Comment (alejandromumo) [reply invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultItem.js:59]: minor: to avoid code duplication consider something like [element variables](https://reactjs.org/docs/conditional-rendering.html#element-variables):

```javascript
const formattedElement = <Formatter result={result} resourceSchema={resourceSchema} property={property}/>;
const columnElement = index > 0 ? formattedElement : <a href="...">{formattedElement}<a/>;

return (
   {columnElement}
)
```

However you might only "shrinken" the `Formatter` element and not the conditional part. 

Imo both options save some space, given that ~50% of the component is a conditional + the `Formatter` element whose instantiation is duplicated.

File: invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultsContainer.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+export const SearchResultsContainer = ({
+  results,
+  columns,
+  displayEdit,
+  displayDelete,
+  actions,
+}) => {
+  const resourceHasActions = displayEdit || displayDelete || !isEmpty(actions);
+
+  const columnWidth = {
```
Comment (alejandromumo) [inline invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultsContainer.js:24]: I am not sure about this approach, it's i) hardcoding field names and ii) ties the front-end rendering to the current state of the backend. E.g. in the future the representation of a page might change and this code won't work anymore. 

However I don't have any suggestion, if anyone else wants to suggest feel free

File: invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultItem.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+  actions: PropTypes.object,
+  updateQueryState: PropTypes.func.isRequired,
+  currentQueryState: PropTypes.object.isRequired,
+  idKeyPath: PropTypes.string.isRequired,
+  resourceSchema: PropTypes.object.isRequired,
+  listUIEndpoint: PropTypes.string.isRequired,
+};
+
+SearchResultItemComponent.defaultProps = {
+  displayDelete: true,
```
Comment (alejandromumo) [inline invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultItem.js:117]: should it be enabled by default?

File: invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultsContainer.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+SearchResultsContainer.propTypes = {
+  results: PropTypes.array.isRequired,
+  columns: PropTypes.array.isRequired,
+  displayEdit: PropTypes.bool,
+  displayDelete: PropTypes.bool,
+  actions: PropTypes.object.isRequired,
+};
+
+SearchResultsContainer.defaultProps = {
+  displayDelete: true,
```
Comment (alejandromumo) [inline invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultsContainer.js:65]: should it be enabled by default? In case something goes wrong while rendering the component, better to be safe and disable things.

File: invenio_pages/records/errors.py [hunk 11]
```diff
@@ -7,28 +7,32 @@
...
-        self.identifier = identifier
-
-    @property
-    def description(self):
-        """Exception's description."""
-        return f"The page identified by {self.identifier} cannot be found."
+        """Initialise error."""
+        super().__init__(
+            _(
+                "The page identified by {identifier} cannot be found.".format(
```
Comment (alejandromumo) [inline invenio_pages/records/errors.py:20]: You can use  `f` notation for formatting, since this is running in python >= 3 . Applies to other occurrences in this PR.

```suggestion
                f"The page identified by {self.identifier} cannot be found."
```

File: invenio_pages/views.py (hunk 5 above)
Comment (alejandromumo) [reply invenio_pages/views.py:64]: `to_obj` was returning `self._record`. Doesn't `to_dict()` return other properties as well? In any case, the equivalaent to your `to_obj` property I think it would be [`RecordItem._obj`](https://github.dev/inveniosoftware/invenio-records-resources/blob/a2ab7ab5feedfef480e41821ee01d347883daacf/invenio_records_resources/services/records/results.py#L59) in case you want to play it safe.

**Review by alejandromumo — COMMENTED** [review]
Thanks for the PR! 

Left some comments.

File: invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultsContainer.js (hunk 8 above)
Comment (Pineirin) [reply invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultsContainer.js:24]: Unless we adjust the width of each column according to the size of the content of all the cells, I don't think there is a better way to define the size of the columns. I also don't think is that problematic to hardcode the field names, as this is not a generic module but an override. Finally, even do it's true that we are tying the front-end rendering to the current-state of the backend, without doing it an going for generic widths like in the base invenio-administration, the column distribution looks wrong.

File: invenio_pages/records/errors.py (hunk 11 above)
Comment (Pineirin) [reply invenio_pages/records/errors.py:20]: I'm not sure about this but I think that using f templating for strings was not interacting properly with flask_babelex, but I cannot remember why.

File: invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultsContainer.js (hunk 8 above)
Comment (Pineirin) [reply invenio_pages/assets/semantic-ui/js/invenio_pages/search/SearchResultsContainer.js:24]: I just saw an alternative in anika's PR so I will go with it to avoid defining here the columnWidths.

File: invenio_pages/services/permissions.py [hunk 12]
```diff
@@ -15,8 +15,8 @@
 class PagesPermissionPolicy(BasePermissionPolicy):
     """Permission policy for pages."""
 
-    can_create = []
+    can_create = [SystemProcess()]
```
Comment (jrcastro2) [inline invenio_pages/services/permissions.py:18]: Minor: why this change? In reallity both (creation and deletion) are disabled, or am I missing something?

**Review by jrcastro2 — COMMENTED** [review]
Nice to see the integration working, good job! 

I left a few comments regarding the possibilty to avoid the duplicated code from invenio-administration. 
To me it looks like the cases that you are trying to solve are generic enough to add them to the core module.
By doing so we would also avoid adding the webpack to the module :smile: 

Feel free to ping me if you have any questions!

File: invenio_pages/services/permissions.py (hunk 12 above)
Comment (Pineirin) [reply invenio_pages/services/permissions.py:18]: Because the create and delete services need to be called by the fixtures of invenio-app-rdm

File: invenio_pages/webpack.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
```
Comment (jrcastro2) [inline invenio_pages/webpack.py:1]: Moderate: This can be removed.

File: setup.cfg [hunk 14]
```diff
@@ -66,6 +66,12 @@ invenio_base.api_blueprints =
     invenio_pages = invenio_pages.views:create_pages_api_bp
 invenio_db.alembic =
     invenio_pages = invenio_pages:alembic
+invenio_administration.views =
+    invenio_pages_list = invenio_pages.administration.views.pages:PageListView
+    invenio_pages_details = invenio_pages.administration.views.pages:PageDetailView
+    invenio_pages_create = invenio_pages.administration.views.pages:PageEditView
+invenio_assets.webpack =
+    invenio_pages = invenio_pages.webpack:theme
```
Comment (jrcastro2) [inline setup.cfg:74]:
```suggestion
```

Moderate: This is not needed anymore.

File: invenio_pages/administration/views/pages.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    display_search = True
+    display_delete = False
+    display_create = False
+    display_edit = True
+
+    item_field_list = {
+        "url": {"text": _("Url"), "order": 1, "width": 1},
+        "title": {"text": _("Title"), "order": 2, "width": 2},
+        "content": {"text": _("Content"), "order": 3, "width": 5, "truncate": 250},
+        "template_name": {"text": _("Template Name"), "order": 4, "width": 3},
```
Comment (kpsherva) [inline invenio_pages/administration/views/pages.py:39]: I think the content of the page and template name are not needed in the search view. It is overloading the user with information. In addition, the content is not really nicely displayed due to html format

File: invenio_pages/config.py [hunk 16]
```diff
@@ -8,6 +8,8 @@
 
 """Page module config."""
 
+from flask_babelex import lazy_gettext as _
```
Comment (kpsherva) [inline invenio_pages/config.py:11]: @utnapischtim having your changes in mind, is this import correct?

File: invenio_pages/config.py [hunk 17]
```diff
@@ -21,3 +23,32 @@
     "THEME_SITENAME",
 ]
 """List of configuration variables accessible in the dynamic pages."""
+
+PAGES_SEARCH = {
+    "facets": [],
+    "sort": ["url", "title", "created", "updated"],
+}
+"""Page search configuration (i.e list of pages)"""
```
Comment (kpsherva) [inline invenio_pages/config.py:31]:
```suggestion
"""Page search configuration (i.e list of pages)."""
```

File: invenio_pages/records/errors.py (hunk 11 above)
Comment (kpsherva) [reply invenio_pages/records/errors.py:20]: it is probably because f-strings are evaluated in runtime, and gettext is gathering all strings for translations before the f-string gets evaluated ? you can confrim that by checking with lazy_gettext: https://pythonhosted.org/Flask-BabelEx/#using-translations

File: invenio_pages/views.py (hunk 5 above)
Comment (kpsherva) [reply invenio_pages/views.py:64]: do we have consensus here ? @ppanero

**Review by kpsherva — COMMENTED** [review]
code LGTM!
only little UI improvements, discussed in person

File: invenio_pages/config.py (hunk 16 above)
Comment (Pineirin) [reply invenio_pages/config.py:11]: @kpsherva Reminder to take this into account after releasing the changes.

**Comment by jrcastro2** [comment]
Tests are failing (PYDOCSTYLE)
